ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpPreprocessedLexer.cs

```diff
@@ -0,0 +1,317 @@
+    private readonly FSharpList<string> myDefinedConstants;
+    private ImmutableQueue<Tuple<FSharpLexerState, string>> myQueue =
+      ImmutableQueue<Tuple<FSharpLexerState, string>>.Empty;
+    private PreprocessorState myState = new PreprocessorState();
+
+    public FSharpPreprocessedLexer(IBuffer buffer, IFSharpPreprocessor preprocessor, FSharpList<string> definedConstants)
+    {
+      myLexer = new FSharpLexer(buffer);
+      myPreprocessor = preprocessor;
+      myDefinedConstants = definedConstants;
+    }
+
+    public FSharpPreprocessedLexer(
+      IBuffer buffer,
+      int startOffset,
+      int endOffset,
+      IFSharpPreprocessor preprocessor,
+      FSharpList<string> definedConstants)
+    {
+      myLexer = new FSharpLexer(buffer, startOffset, endOffset);
+      myPreprocessor = preprocessor;
+      myDefinedConstants = definedConstants;
+    }
+
+    public void Start()
+    {
+      myLexer.Start();
+      currTokenType = null;
+    }
+
+    private void Restore(Tuple<FSharpLexerState, string> stackElem)
+    {
+      myLexer.CurrentPosition = stackElem.Item1;
+    }
+
+    private FSharpLexerState DeadCodeToken(FSharpLexerState state, int dumpStart)
+    {
+      state.currTokenType = FSharpTokenType.DEAD_CODE;
+      state.yy_buffer_start = dumpStart;
+      state.yy_buffer_end = myLexer.TokenStart;
+      state.yy_buffer_index = myLexer.TokenStart;
+      return state;
+    }
+
+    private TokenNodeType PreprocessInactiveBranch()
+    {
+      switch (myLexer.TokenType)
+      {
+        case var token when token == FSharpTokenType.PP_IF_SECTION:
+          myState.InIfBlock(false, false);
+          return PreprocessInactiveLine();
+        case var token when token == FSharpTokenType.PP_ENDIF:
+          myState.FromIfBlock();
+          return myState.Condition ? PreprocessLine() : PreprocessInactiveLine();
+        case var token when token == FSharpTokenType.PP_ELSE_SECTION:
+          myState.SwitchBranch();
+          return myState.Condition ? PreprocessLine() : PreprocessInactiveLine();
+        default:
+          return PreprocessInactiveLine();
+      }
+    }
+
+    private bool Skip(TokenNodeType token)
+    {
+      return token == FSharpTokenType.WHITESPACE ||
+             token == FSharpTokenType.PP_IF_SECTION ||
+             token == FSharpTokenType.NEW_LINE;
+    }
+
+    private TokenNodeType PreprocessIfSection()
+    {
+      TokenNodeType tokenType;
+      do {
+        tokenType = myLexer.TokenType;
+        myQueue = myQueue.Enqueue(new Tuple<FSharpLexerState, string>(
+          myLexer.CurrentPosition,
+          myLexer.GetCurrTokenText())
+        );
+        myLexer.Advance();
+      } while (!(tokenType == FSharpTokenType.NEW_LINE || tokenType == null));
+      var list = ListModule.OfSeq(myQueue
+        .Where(x => !Skip(x.Item1.currTokenType))
+        .Select(x => new Tuple<TokenNodeType, string>(x.Item1.currTokenType, x.Item2)));
```

Comment:
I'd try avoid using Linq and tuples here, it looks like too many allocations for a preprocessor.